Incr1 + incr2 + incr3

zueve opened the pull request.

zueve committed:
First inr

zueve committed:
Add layers

zueve committed:
Upd pkg name

zueve committed:
Add tests

zueve committed:
Some fixes

zueve committed:
Inct #3: add chi

zueve committed:
Fix entrypoint

zueve committed:
Fix entrypoint v2

zueve committed:
Support plain/text type

zueve committed:
Support plain/text type

zueve committed:
Change response format

zueve committed:
Switch on chi.UrlParam

zueve committed:
Inc3: fix statictest

iTiky commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,37 @@
+	c.counter++
+	key := strconv.Itoa(c.counter)
+	c.links[key] = url
+	return key
```
Лучше перед return ставить пустую строку, сильно улучшает читабельность кода (видны точки выхода из функции "издалека").

iTiky commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,37 @@
+}
+
+func (c *Storage) Get(key string) (string, error) {
+	url, ok := c.links[key]
```
`map` не потокобезопасны и для чтения: если кто-то пишет, а тут пришел читатель - паника. Я бы посоветовал `sync.RWMutex` для этих целей.

iTiky commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,37 @@
+func (c *Storage) Get(key string) (string, error) {
+	url, ok := c.links[key]
+	if !ok {
+		return "", errors.New("key not exist")
```
Я бы изменил прототип на `Get(key string) (*string, error)`. С точки зрения БД, ключ не найден это не ошибка. Если не найдет будет возврат `(nil, nil)`, который удобно обработать на уровне логики.

iTiky commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,37 @@
+func New() Storage {
+	return Storage{
+		counter: 1,
+		links:   map[string]string{},
```
Лучше создавать `map` / `slice` через `make`, если в этом же scope (пакет в данном случае), будет запись.

iTiky commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,37 @@
+)
+
+type Storage struct {
+	mtx     sync.Mutex
```
Такое принято оформлять через struct embedding:
```go
type Storage struct {
    sync.RWMutex

    links ...
}
```

iTiky commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,37 @@
+	return url, nil
+}
+
+func New() Storage {
```
Конструктор лучше перенести выше (сразу после объявления типа). Так легче читать код другим разрабам: код смотрят сверху вниз и так будут сразу ответы на вопросы "что, как создать, как использовать".

iTiky commented on internal/services/services.go:
```diff
@@ -0,0 +1,24 @@
+package services
+
+type IStorage interface {
```
1. В Go принято иначе именовать интерфейсы: Writer, Storer, Updater и т.д. Префикс обычно не ставят. В данном случае я бы переименовал в `StorerExpected`, т.к. тут объявлено ожидаемое поведение зависимости.
2. Интерфейсы, обычно, оформляют отдельным файлом для упрощения навигации по коду + так легче генерить моки (например через gomock).

iTiky commented on internal/services/services.go:
```diff
@@ -0,0 +1,24 @@
+	return s.storage.Get(key)
+}
+
+func New(storage IStorage) Service {
```
`IStorage` может быть `nil`, хорошо бы добавить на это проверку и вернуть ошибку, если конструктор не может создать валидный объект из входов.

iTiky commented on internal/server/server.go:
```diff
@@ -0,0 +1,74 @@
+	port       int
+}
+
+func New(service services.Service, serviceURL string, port int) Server {
```
Очень "большой" конструктор, тяжело его будет поддерживать. Почитай про паттерн опций, это общепринятый подход. Например тут: https://medium.com/star-gazers/go-options-pattern-da49185a2526

iTiky commented on internal/server/server.go:
```diff
@@ -0,0 +1,74 @@
+func (s *Server) createRedirect(w http.ResponseWriter, r *http.Request) {
+	headerContentType := r.Header.Get("Content-Type")
+	w.Header().Set("content-type", "text/plain")
+	var url = ""
```
Такой стиль не часто вижу. Есть два "стандартных" варианта:
1. url := ""
2. var url string
Когда 1-ый, а когда 2-ой?

1-ый, когда мы объявили переменную и задали ей значение, которое можно использовать далее по коду.
2-ой, когда нам не важно текущее значение и так мы показываем, что инициализация будет дальше по коду.

Твой вариант плох тем, что строка by default `== ""` и не понятен смысл строки.

iTiky commented on internal/server/server.go:
```diff
@@ -0,0 +1,74 @@
+	headerContentType := r.Header.Get("Content-Type")
+	w.Header().Set("content-type", "text/plain")
+	var url = ""
+	if headerContentType == "application/x-www-form-urlencoded" {
```
if - else if - else if - else. Лучше избегать этого (справедливо не только в Go). Это сильно увеличивает когнитивную нагрузку на читателя кода. Почти всегда можно избавить от else, в данном случае:
```go
switch headerContentType {
    case "application/x-www-form-urlencoded":
    default:
}
```

iTiky commented on internal/server/server.go:
```diff
@@ -0,0 +1,74 @@
+	} else if headerContentType == "text/plain; charset=utf-8" {
+		urlBytes, err := io.ReadAll(r.Body)
+		if err != nil {
+			w.WriteHeader(http.StatusUnsupportedMediaType)
```
Если чтение провалилось, это скорее 500 и выход.

iTiky commented on internal/server/server.go:
```diff
@@ -0,0 +1,74 @@
+		url = strings.TrimSuffix(string(urlBytes), "\n")
+	} else {
+		w.WriteHeader(http.StatusUnsupportedMediaType)
+		fmt.Println("invalid ContentType")
```
Очень странные консольные логи. Лучше либо их убрать, либо использовать "большой" логгер с контекстом (кто запросил, какой ендпоинт, что пошло не так).

iTiky commented on internal/server/server.go:
```diff
@@ -0,0 +1,74 @@
+	url, err := s.service.GetURLByKey(key)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Println("invalid key", key)
```
Тут аналогично. Может лучше эту ошибку пользователю отдать в теле ответа?

iTiky commented on cmd/shortener/main.go:
```diff
@@ -1,3 +1,15 @@
+)
+
+func main() {
+	var storageTest = storage.New()
```
-> `storageTest := storage.New()`, такой стиль не принят в Го сообществе

iTiky commented on cmd/shortener/main.go:
```diff
@@ -1,3 +1,15 @@
+
+func main() {
+	var storageTest = storage.New()
+	var serviceTest = services.New(&storageTest)
```
Почему здесь передаётся указатель? `services.New` принимает интерфейс, т.е. можно передать как `storageTest`, так и `&storageTest`. Нам нужен указатель, т.к. storage - mutable объект. В таком случае конструктор `storage.New` должен возвращать указатель. Конструктор должен всегда возвращать валидный и готовый к использованию объект (или ошибку, если не смог его построить). В данном случае тут есть какие-то доп. действия над объект - лучше этого избегать.

iTiky commented on cmd/shortener/main.go:
```diff
@@ -1,3 +1,15 @@
+	var serviceTest = services.New(&storageTest)
+	var serverTest = server.New(serviceTest, "http://localhost:8080", 8080)
+
+	serverTest.Run()
```
Нет graceful shutdown: завершение работы сервера при получении SIGINT сигнала ОС.

iTiky reviewed (changes requested):
Хорошее разбиение на слои. Нужно исправить лок storage и добавить graceful shutdown. Я бы ещё посоветовал начать писать тесты (свои), это крайне полезно.

zueve committed:
Fix style

zueve committed:
Fix style 2

zueve committed:
Fix: lock map on read

zueve committed:
Add config

zueve committed:
Incr3: add graceful shutdown

zueve committed:
incr3: upd Readme

zueve committed:
incr4: add tests

zueve committed:
Incr4: small refactoring

hasansino commented on cmd/shortener/main.go:
```diff
@@ -1,3 +1,39 @@
+)
+
+func main() {
+	storageVar := storage.New()
```
no need to add `Var` to variables, you can name is `s` and it will be easy to read in such small file

hasansino commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,43 @@
+type Storage struct {
+	sync.RWMutex
+	links   map[string]string
+	counter int
```
Можно обойтись без счетчика и мапы.
Сделай слайс []string и добавляй значения туда.
ID запроса это индекс+1.

hasansino commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,43 @@
+
+	url, ok := c.links[key]
+	if !ok {
+		return "", errors.New("key not exist")
```
Вынеси `errors.New("key not exist")` как переменную, 

var ErrKeyNotFound = errors.New("key not exist")
...
return "", ErrKeyNotFound

Так ты сможешь проверить тип ошибки и выдать 404 а не 400

hasansino commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,43 @@
+package storage
```
Если интересно, экстра задачка - этот код возможно написать без мютекса.
Еще посложнее - вообще без пакета sync.
При разных подходах можешь написать бенчмарк (https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go) что-бы понять как синхронизация влияет на производительность.

hasansino commented on internal/config/config.go:
```diff
@@ -0,0 +1,33 @@
+
+type ContextOption func(*Context)
+
+type Context struct {
```
Пакет config -  значит тут должен быть ... конфиг)
Менеджмент контекстов это не задача конфига.

hasansino commented on internal/config/config.go:
```diff
@@ -0,0 +1,33 @@
+	return &uCtx
+}
+
+func WithPort(port int) ContextOption {
```
В ContextOption тут нет смысла - поля и так публичные.
Обычно такой конфигурацией скрывают сложную логику настройки.

zueve committed:
Incr4: refactoring

zueve committed:
incr5: support config from env

zueve committed:
incr5: rename config/config -> config/context

zueve committed:
incr5: rename config/config -> config/context

zueve committed:
incr6: support file storage

zueve committed:
incr6: few fixes

zueve committed:
incr6: fix test in pipeline

zueve committed:
incr7: support cmd flags

zueve committed:
fix naming

zueve committed:
incr8: add compression

zueve committed:
incr8: fix reader

zueve committed:
incr8: support "application/x-gzip:

zueve committed:
incr8: support "application/x-gzip:

zueve committed:
incr9: WIP

incr8: upd storage & services

incr9: add new url /user/urls

zueve committed:
incr8: fix options

zueve committed:
incr8: fixes after code review

incr8: fix forward errors

incr8: rename PersistentStorageExpected

incr8: fix forward error

incr8: fix temt file

incr8: add untracked files

incr8: fix cm flags

incr8: fixes after 1-2-1

incr8: fix tests

zueve committed:
incr9: fix tests

zueve committed:
incr10: db ping

zueve committed:
Merge pull request #6 from zueve/incr8

Sprint 2: Incr4 - Incr8

zueve committed:
incr10: fix tests

zueve committed:
sprint3: fix set cookie

zueve committed:
sprint3: fix status

zueve committed:
sprint3: add context to ping

zueve committed:
sprint3: fix status

zueve committed:
sprint3: fix timeout

zueve committed:
sprint3: fix connection db

zueve committed:
incr11: add db

zueve committed:
sprint3: run tests

zueve committed:
sprint3: refactoring

zueve committed:
sprint3: run migrations on startup

zueve committed:
sprint3: incr12

zueve committed:
sprint3: incr12 - support transactions

zueve committed:
sprint3: incr13

zueve committed:
sprint3: incr12 add context logger

zueve committed:
sprint3: fix interfaces
